packages/services/tokens/src/multi-tier-storage.ts

```diff
@@ -0,0 +1,395 @@
+      });
+      await db.destroy();
+      // Wait for Redis to finish all the pending operations
+      await redis.quit();
+      touch.dispose();
+    },
+    isReady: atomic(async () => {
+      if (redis.status === 'ready' || redis.status === 'reconnecting') {
+        return db.isReady();
+      }
+      return false;
+    }),
+    async readTarget(target) {
+      const tokens = await db.getTokens({ target });
+      return tokens.map(transformToken);
+    },
+    async readToken(hashedToken, maskedToken) {
+      const status: LRUCache.Status<CacheEntry> = {};
+      const context = { maskedToken, source: 'in-memory' };
+      const data = await cache.fetch(hashedToken, {
+        context,
+        status,
+      });
+
+      if (status.fetch) {
+        recordCacheRead(status.fetch);
+      } else {
+        serverLogger.warn('Status of the fetch is missing');
+      }
+
+      if (!data) {
+        // Looked up in all layers, and the token is not found
+        return null;
+      }
+
+      if (data === 'not-found') {
+        return null;
+      }
+
+      touch.schedule(hashedToken);
+      return data;
+    },
+    writeToken: tracker.wrap(async item => {
+      const result = await db.createToken(item);
+
+      // We don't want to write to in-memory cache,
+      // as the token may not be used immediately, or at all.
+      // We write to Redis, so in case the token is used,
+      // we reuse it for the in-memory cache, without hitting the DB.
+      const hashedToken = result.token;
+      const cacheEntry = transformToken(result);
+
+      // Write to Redis gracefully. If it fails, we log the error, but we don't throw.
+      // The token won't be in Redis cache, but it will be possible to retrieve it from the DB.
+      // It will affect performance slightly, but it won't break the system.
+      try {
+        await setInRedis(redis, hashedToken, cacheEntry);
+      } catch (error) {
+        handleStorageError({
+          logger: serverLogger,
+          error,
+          logMsg: 'Failed to write token to Redis, but it was created in the Db',
+          tier: 'redis',
+          action: 'set',
+        });
+      }
+
+      return cacheEntry;
+    }),
+    deleteToken: tracker.wrap(async hashedToken => {
+      await db.deleteToken({
+        token: hashedToken,
+        async postDeletionTransaction() {
+          // delete from Redis when the token is deleted from the DB
+          await redis.del([generateRedisKey(hashedToken), generateStaleRedisKey(hashedToken)]);
+          // only then delete from the in-memory cache.
+          // The other replicas will purge the token from
+          // their own in-memory caches on their own pace (ttl)
+          cache.delete(hashedToken);
+        },
+      });
+    }),
+    invalidateTokens: tracker.wrap(async hashedTokens => {
+      if (hashedTokens.length === 0) {
+        return;
+      }
+
+      await redis.del(
```

Comment:
Should we run these in parallel? or at least make sure to invalid cache even if Redis failed? 🤔 
I mean, if this function is called, we know that we need to clean the token for sure.

Reply:
Redis **needs** to be invalidated, because it stores data for much much longer, in-memory cache **can** as the token will expire in less than 5 minutes